agent/credentials/providers/rotating_shared_credentials_provider.go

```diff
@@ -0,0 +1,65 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package providers
+
+import (
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/cihub/seelog"
+)
+
+const (
+	// DefaultRotationInterval is how frequently to expire and re-retrieve the credentials from file.
+	DefaultRotationInterval = time.Minute
+	// DefaultFilename is the default location of the credentials file within the container.
+	DefaultFilename = "/rotatingcreds/credentials"
+	// RotatingSharedCredentialsProviderName is the name of this provider
+	RotatingSharedCredentialsProviderName = "RotatingSharedCredentialsProvider"
+)
+
+// RotatingSharedCredentialsProvider is a provider that retrieves credentials via the
+// shared credentials provider, and adds the functionality of expiring and re-retrieving
+// those credentials from the file.
+type RotatingSharedCredentialsProvider struct {
+	credentials.Expiry
+
+	RotationInterval          time.Duration
+	sharedCredentialsProvider *credentials.SharedCredentialsProvider
+}
+
+// NewRotatingSharedCredentials returns a rotating shared credentials provider
+// with default values set.
+func NewRotatingSharedCredentialsProvider() *RotatingSharedCredentialsProvider {
+	return &RotatingSharedCredentialsProvider{
+		RotationInterval: DefaultRotationInterval,
+		sharedCredentialsProvider: &credentials.SharedCredentialsProvider{
+			Filename: DefaultFilename,
+			Profile:  "default",
+		},
+	}
+}
+
+// Retrieve will use the given filename and profile and retrieve AWS credentials.
+func (p *RotatingSharedCredentialsProvider) Retrieve() (credentials.Value, error) {
+	v, err := p.sharedCredentialsProvider.Retrieve()
+	v.ProviderName = RotatingSharedCredentialsProviderName
+	if err != nil {
+		return v, err
+	}
+	p.SetExpiration(time.Now().Add(p.RotationInterval), 0)
+	seelog.Infof("Successfully got instance credentials from file %s. Access Key ID XXXX%s from %s",
+		p.sharedCredentialsProvider.Filename, v.AccessKeyID[len(v.AccessKeyID)-4:], v.ProviderName)
```

Comment:
this will panic if someone puts an invalid key id with less than 4 characters

Reply:
good catch, will fix